framework/calico_framework.py

```diff
@@ -0,0 +1,673 @@
+                self.cpus += resource.scalar.value
+            elif resource.name == "mem":
+                self.mem += resource.scalar.value
+            elif resource.name == "ports":
+                port_range = resource.ranges.range[0]
+                self.port = randint(port_range.begin, port_range.end)
+
+    def __eq__(self, other):
+        return self.offer.slave_id == other.offer.slave_id
+
+    def __hash__(self):
+        return hash(self.offer.slave_id.value)
+
+    def num_tasks_launchable(self):
+        return int(math.floor(min(self.cpus / TASK_CPUS, self.mem / TASK_MEM)))
+
+    def __repr__(self):
+        return "Offer(id=%s, tasks=%d, slave=%s)" % (
+            self.offer_id[-5:], self.num_tasks_launchable(),
+            self.offer.slave_id.value[-5:])
+
+
+class TestScheduler(mesos.interface.Scheduler):
+    def __init__(self, implicit_acknowledgements):
+        self.implicitAcknowledgements = implicit_acknowledgements
+        """
+        Flag to disable the requirement that the Executor responds to ACK
+        messages
+        """
+
+        self.tests = []
+        """
+        A collection of TestCases this scheduler will run through should deploy.
+        """
+
+        self.test_by_slave_id = {}
+        """
+        TODO: describame
+        """
+
+    def all_tasks(self):
+        """
+        Get all tasks across all tests.
+        """
+        for test in self.tests:
+            for task in test.tasks:
+                yield task
+
+    def all_tests_complete(self):
+        return len([test for test in self.tests if
+                    test.complete()]) == len(self.tests)
+
+    def unreserve_offers_by(self, test):
+        for slave_id, reserving_test in self.test_by_slave_id.iteritems():
+            if test == reserving_test:
+                del(self.test_by_slave_id[slave_id])
+
+
+    def kill_test(self, test, msg=None):
+        # Mark the test as done, and explain why
+        test.killed = True
+        test.additional_info = msg
+        if msg:
+            _log.error("Test Killed: %s. Reason: %s", test, msg)
+
+        # Unreserve any offers reserved by this test
+        self.unreserve_offers_by(test)
+
+        # If there's no other tests, we're done here
+        if self.all_tests_complete():
+            self.report_results_and_exit()
+
+    def registered(self, driver, frameworkId, masterInfo):
+        """
+        Callback used when the framework is succesfully registered.
+        """
+        _log.info("REGISTERED: with framework ID %s", frameworkId.value)
+
+    def resourceOffers(self, driver, offers):
+        """
+        Triggered when the framework is offered resources by mesos.
+        """
+        unreserved_offers_by_slave_id = {}
+
+        # Send new offers to tasks which reserved them, or add to unreserved
+        # offers pool
+        for offer in offers:
+            new_offer = Offer(offer)
+            if new_offer.slave_id in self.test_by_slave_id:
```

Comment:
Nit:  The usual Python thing to do here is a try/except.  That's because evaluating `a in b` means generating a hash of `a`, which then has to be done a second time to execute `b[a].do_a_thing()`.